### PR TITLE
offsets/liboffsets cleanup

### PR DESCRIPTION
**Pull request by jbrockmendel**
working on making DateOffset immutable enough unrelated changes added up to merit separating them out into this PR.  No logic is affected.

### CODE REVIEW

File: pandas/_libs/tslibs/offsets.pyx [hunk 1]
```diff
@@ -418,6 +510,31 @@ cdef inline int month_add_months(pandas_datetimestruct dts, int months) nogil:
     return 12 if new_month == 0 else new_month
 
 
+def shift_day(datetime other, int days):
+    """
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx]: def?

Comment (jbrockmendel) [reply pandas/_libs/tslibs/offsets.pyx]: At the moment it’s only called from python-land.  When that changes we’ll cpdef it.

File: pandas/_libs/tslibs/offsets.pyx [hunk 2]
```diff
@@ -418,6 +509,31 @@ cdef inline int month_add_months(pandas_datetimestruct dts, int months) nogil:
...
+    -------
+    shifted: datetime or Timestamp
+    """
+    if other.tzinfo is None:
+        return other + timedelta(days=days)
+
+    tz = other.tzinfo
+    naive = other.replace(tzinfo=None)
+    shifted = naive + timedelta(days=days)
+    return _localize_pydatetime(shifted, tz)
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx]: so I think we ought to de-privatize _localize_pydatetime (do this now) then and in future cimport this

File: pandas/_libs/tslibs/offsets.pyx [hunk 3]
```diff
@@ -95,6 +101,27 @@ def as_datetime(obj):
     return obj
 
 
+cpdef datetime as_timestamp(obj):
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx]: should be in timestamps.pyx

**Review by jreback — CHANGES_REQUESTED** [review]

**Pull request closed without merging**